### PR TITLE
fix(gateguard): match destructive SQL passed to a database client in a quoted argument

### LINKED ISSUE #3024: gateguard: DESTRUCTIVE_SQL_DD is unreachable for quoted SQL — stripQuotedStrings() runs before the SQL arm
# `DESTRUCTIVE_SQL_DD` never sees quoted SQL, so every realistic SQL invocation is allowed

**Summary:** `stripQuotedStrings()` is applied to the command *before* the `DESTRUCTIVE_SQL_DD` arm tests it, at both call sites. Since SQL is almost always passed to a client inside a quoted `-c` / `-e` argument, the destructive SQL patterns are unreachable in practice.

**Affected version:** `2.2.1`, commit `5064474d4d762dc9640234a41617cccb79185cec`, installed via the `ecc@ecc` marketplace plugin. Installed `scripts/hooks/gateguard-fact-force.js` is 1347 lines.

## Reproduction

Save as `repro.js` and run `node repro.js <plugin>`, where `<plugin>` is the installed plugin directory.

```js
const os = require('os'), fs = require('fs'), path = require('path');

// Isolate hook state so the repro cannot touch a real ~/.gateguard directory.
process.env.GATEGUARD_STATE_DIR = fs.mkdtempSync(path.join(os.tmpdir(), 'ggrepro-'));

const { run } = require(process.argv[2] + '/scripts/hooks/gateguard-fact-force.js');

function decide(command) {
  const input = { tool_name: 'Bash', tool_input: { command }, session_id: 'repro' };
  return run(input) === input ? 'ALLOWED' : 'DENIED';
}

decide('printf ready'); // prime: the first Bash call trips the separate routine gate

for (const command of [
  'psql -c "drop table users"',
  "psql -c 'truncate audit_log'",
  'mysql -e "delete from sessions"',
  'psql -c drop table users',
]) console.log(decide(command).padEnd(8), command);
```

## Observed results

| Command | Decision |
| --- | --- |
| `psql -c "drop table users"` | **ALLOWED** |
| `psql -c 'truncate audit_log'` | **ALLOWED** |
| `mysql -e "delete from sessions"` | **ALLOWED** |
| `psql -c drop table users` (unquoted) | DENIED |

Only the unquoted spelling is caught — and that spelling is not valid for the client, since the shell would split it into separate arguments.

## Root cause

Line 60 defines the arm:

```js
const DESTRUCTIVE_SQL_DD = /\b(drop\s+table|delete\s+from|truncate|dd\s+if=)\b/i;
```

Both call sites strip quoted content first — lines 702-703:

```js
const flattened = explodeSubshells(stripQuotedStrings(executable));
if (DESTRUCTIVE_SQL_DD.test(flattened)) return true;
```

and lines 729-730:

```js
const stripped = stripQuotedStrings(segment);
if (DESTRUCTIVE_SQL_DD.test(stripped)) return true;
```

The stripping is deliberate, and the comment above line 60 explains why: it stops a commit message that mentions `drop table` from tripping the gate. That rationale is sound for the false-positive direction, but it also guarantees the SQL patterns can never match a real invocation, because a real invocation must quote the statement.

So the three SQL patterns (`drop table`, `delete from`, `truncate`) are effectively dead code. `dd if=` on the same line is separately unreachable for real device paths — that one is [#2642](https://github.com/affaan-m/ECC/issues/2642), which I have commented on with a 2.2.1 confirmation.

## Impact

An operator who trusts GateGuard to gate destructive commands gets no speed bump before an agent drops or truncates a table, or deletes every row. The failure is silent: the command is allowed with no indication a pattern was intended to cover it.

## Suggested fix

Keep the quote-stripping for the generic scan, but scan *inside* the quoted argument when the executable is a known SQL client. Roughly: if the resolved command basename is one of `psql`, `mysql`, `mariadb`, `sqlite3`, `mongosh`, `clickhouse-client`, etc., extract the argument to `-c` / `-e` / `--command` / `--execute` and test the SQL patterns against that string specifically.

That preserves the commit-message protection — `git commit -m "drop table cleanup"` still has no SQL client in the executable position — while making the patterns reachable where they matter.

Regression cases worth pinning, all of which should deny:

```
psql -c "drop table users"
psql --command='truncate audit_log'
mysql -e "delete from sessions"
sqlite3 app.db "drop table users"
```

and this should continue to allow:

```
git commit -m "refactor: drop table indirection"
```

## Notes

Searched the tracker for `psql`, `sql`, `drop table`, `quoted`, and `strip` before filing. The nearest existing reports are [#2886](https://github.com/affaan-m/ECC/issues/2886) (detector matches heredoc body text) and [#3023](https://github.com/affaan-m/ECC/issues/3023) (block-no-verify false positive on a quoted argument) — both over-matching bugs, i.e. the opposite direction from this one.

Filed as a public issue rather than a private advisory to match the maintainers' handling of the same class of defect (#2642, #2291, #2886 are all public). Happy to move it to a private advisory if you would rather triage destructive-gate bypasses that way.

### PR DESCRIPTION
**Pull request by L4XB**
## What Changed
`scripts/hooks/gateguard-fact-force.js`: the quote-aware pass now tests the `DESTRUCTIVE_SQL_DD` phrases (`drop table`, `delete from`, `truncate`) against the unquoted argument text of a segment whose executable is a known database client (`psql`, `pgcli`, `mysql`, `mariadb`, `mycli`, `sqlite3`, `duckdb`, `usql`, `sqlcmd`, `clickhouse-client`, `cockroach`, `mongosh`, `cqlsh`, `trino`, `presto`, `beeline`, ...). The generic quote-stripped scan is unchanged, so a commit message that mentions `drop table` still passes. Because the check lives in `isDestructiveQuoteAware`, it also applies behind `sh -c` and in chained segments.

Tests added to `tests/hooks/gateguard-fact-force.test.js`:
- deny: `psql -c "drop table users"`, `psql -c 'truncate audit_log'`, `psql --command='truncate audit_log'`, `mysql -e "delete from sessions"`, `sqlite3 app.db "drop table users"`, `/usr/local/bin/psql -h db -U app -c "DROP TABLE users"`, `sh -c "psql -c 'drop table users'"`, `echo migrating && mysql -e "truncate sessions"`
- allow: `psql -c "select count(*) from users"`, `psql -c "delete_from_queue()"`, `git commit -m "refactor: drop table indirection"`, `echo "psql -c 'drop table users'"`

## Why This Change
Fixes #3024. `stripQuotedStrings()` ran before the `DESTRUCTIVE_SQL_DD` arm at both call sites, and a real SQL invocation must quote the statement, so the SQL phrases were unreachable: `psql -c "drop table users"` and `mysql -e "delete from sessions"` were allowed, only the unquoted spelling (which the shell would split into separate arguments) was caught. The stripping itself is right for the generic scan (it is what keeps commit messages from tripping the gate), so the fix scopes the unquoted scan to database clients in argv0 position instead of removing it.

## Testing Done
- [x] Manual testing completed (the reproducer from the issue: the four quoted commands now deny, the commit message still allows)
- [x] Automated tests pass locally (`node tests/hooks/gateguard-fact-force.test.js`: 199 passed; the two new deny tests fail on `main`)
- [x] Edge cases considered and tested (client with a path prefix and extra flags, `--command=` inline form, positional statement for `sqlite3`, shell wrapper, chained segment, `select`, identifier containing a keyword, `echo` of a SQL command string)

## Type of Change
- [x] `fix:` Bug fix

## Security & Quality Checklist
- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly (none changed)
- [x] Shell scripts pass shellcheck (if applicable) (none changed)
- [x] Pre-commit hooks pass locally (if configured) (`npx eslint` on the changed files)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

### CODE REVIEW

File: scripts/hooks/gateguard-fact-force.js [hunk 1]
```diff
@@ -376,6 +404,15 @@ function isDestructiveQuoteAware(raw, depth = 0) {
     if (isDestructiveGit(tokens)) return true;
     if (isDestructiveFindExec(tokens.join(' '))) return true;
     const base = commandBasename(tokens[0]);
+    // SQL passed to a database client lives inside a quoted argument, which
+    // the generic scan strips on purpose (a commit message that mentions
+    // "drop table" must not trip the gate). The tokens here carry the
+    // unquoted argument text, so test the SQL phrases on them when the
+    // executable is a known SQL client — `git commit -m "drop table ..."`
+    // still has no SQL client in argv0.
+    if (SQL_CLIENTS.has(base) && DESTRUCTIVE_SQL_DD.test(tokens.slice(1).join(' '))) {
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Done in the follow-up commit: `stripExecutionPrefixes` drops `env` (with `NAME=value` assignments and `-u NAME`), `sudo`/`doas` (with their value options such as `-u user`), `command`, `exec`, `nice -n N`, `nohup`, `time`, `timeout DURATION`, `stdbuf`, `ionice`, `setsid`, `unshare`, `busybox`, `runuser`, `chroot` and `taskset` before argv0 is classified. Tests: `env PGHOST=db psql -c "drop table users"`, `command mysql -e "delete from sessions"`, `sudo -u postgres psql ...`, `nice -n 10 psql ...`, `timeout 30 mysql ...`, stacked `sudo env ...`.

[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Fixed: execution prefixes (`env` incl. assignments, `sudo -u user`, `command`, `nice`, `timeout`, ...) are stripped with their options before the SQL-client check, so both examples are denied now; regression tests added for each prefix and for stacked prefixes.

[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Fixed: the shell-wrapper branch now looks for a short-option cluster containing `c` (`-c`, `-lc`, `-ec`, `-xc`) instead of a standalone `-c`, so `bash -lc 'psql -c "drop table users"'` and `sh -ec 'mysql -e "delete from sessions"'` recurse into the command string and are denied. Both are covered by tests.

[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Addressed as well. Both examples were actually denied by the pre-existing quote-stripped phrase scan (`\btruncate\b` matched the unquoted filename), not only by the new client-argument check, so filename-like words (a path separator or a file extension, `if=` excepted so `dd if=disk.img` still matches) are now skipped in both places. `sqlite3 truncate.db "select 1"`, `psql -f truncate.sql` and `psql --file=./sql/truncate.sql` are allowed; `truncate -s 0 app.log` and real SQL text are still denied.

File: scripts/hooks/gateguard-fact-force.js [hunk 2]
```diff
@@ -359,6 +359,140 @@ function quoteAwareSegments(input) {
...
 const SHELL_WRAPPERS = new Set(['sh', 'bash', 'zsh', 'dash', 'ksh']);
 
+// Programs that only run the command following them. They are stripped
+// (together with their own options and `NAME=value` assignments) before
+// argv0 is classified, so `env PGHOST=db psql -c "..."`, `sudo -u postgres
+// psql -c "..."` and `timeout 30 mysql -e "..."` are treated like the bare
+// client. `valueOptions` consume the next token; `positionals` are leading
+// operands of the prefix itself (`timeout DURATION`, `chroot NEWROOT`).
+const EXECUTION_PREFIXES = {
+  env: { assignments: true, valueOptions: ['-u', '--unset', '-C', '--chdir', '-S', '--split-string'] },
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Done: `env -S <string>` / `--split-string=<string>` (separate or attached) payloads are collected by `stripExecutionPrefixes` and scanned recursively as command lines, so `env -S 'psql -c "drop table users"'` is denied. Tests for both forms.

File: scripts/hooks/gateguard-fact-force.js [hunk 3]
```diff
@@ -359,6 +359,140 @@ function quoteAwareSegments(input) {
...
+      '-u', '--user', '-g', '--group', '-C', '--close-from', '-D', '--chdir', '-h', '--host',
+      '-p', '--prompt', '-r', '--role', '-t', '--type', '-T', '--command-timeout', '-U', '--other-user',
+    ],
+  },
+  doas: { valueOptions: ['-u', '-C'] },
+  command: {},
+  exec: { valueOptions: ['-a'] },
+  nice: { valueOptions: ['-n', '--adjustment'] },
+  nohup: {},
+  time: {},
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Done: `time` now models `-f/--format` and `-o/--output` as value options, so `env time -f "%E" psql -c "drop table users"` and `/usr/bin/time -o /tmp/t.log psql -c "truncate audit_log"` are denied. Both covered by tests.

File: scripts/hooks/gateguard-fact-force.js [hunk 4]
```diff
@@ -359,6 +359,140 @@ function quoteAwareSegments(input) {
...
+/**
+ * Drop leading execution prefixes (`env`, `sudo`, `nice`, ...) with their
+ * options so the real command sits at index 0.
+ *
+ * @param {string[]} tokens
+ * @returns {string[]}
+ */
+function stripExecutionPrefixes(tokens) {
+  let rest = tokens;
+  for (let guard = 0; guard < 8 && rest.length > 0; guard++) {
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Done: the loop now runs until no recognised prefix remains (each round drops at least one token, so it terminates). Regression test with twelve stacked `env` prefixes.

File: scripts/hooks/gateguard-fact-force.js [hunk 5]
```diff
@@ -359,6 +359,159 @@ function quoteAwareSegments(input) {
...
+ * @returns {{ command: string[], commandStrings: string[] }}
+ */
+function stripExecutionPrefixes(tokens) {
+  let rest = tokens;
+  const commandStrings = [];
+  while (rest.length > 0) {
+    const spec = EXECUTION_PREFIXES[commandBasename(rest[0])];
+    if (!spec) break;
+    const valueOptions = new Set(spec.valueOptions || []);
+    const commandStringOptions = spec.commandStringOptions || [];
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js:423]: Done: `runuser` now lists `-c`, `--command` and `--session-command` as command-string options (separate and `--command=` forms), so the payload is scanned recursively. Tests for all three forms.

File: scripts/hooks/gateguard-fact-force.js [hunk 6]
```diff
@@ -359,6 +359,159 @@ function quoteAwareSegments(input) {
...
+      if (token.startsWith('-') && token.length > 1) {
+        const inline = commandStringOptions.find(
+          opt => token.startsWith(opt) && token.length > opt.length && (opt.length === 2 || token.charAt(opt.length) === '=')
+        );
+        if (inline) {
+          commandStrings.push(token.slice(inline.length + (opt => (opt.length === 2 ? 0 : 1))(inline)));
+          i++;
+        } else if (commandStringOptions.includes(token)) {
+          if (rest[i + 1] !== undefined) commandStrings.push(rest[i + 1]);
+          i += 2;
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Done: `commandStrings` is now rebuilt with spread on each addition instead of `push`.

File: scripts/hooks/gateguard-fact-force.js [hunk 7]
```diff
@@ -359,6 +359,164 @@ function quoteAwareSegments(input) {
...
+  nohup: {},
+  time: { valueOptions: ['-f', '--format', '-o', '--output'] },
+  timeout: { valueOptions: ['-s', '--signal', '-k', '--kill-after'], positionals: 1 },
+  stdbuf: { valueOptions: ['-i', '-o', '-e', '--input', '--output', '--error'] },
+  ionice: { valueOptions: ['-c', '--class', '-n', '--classdata', '-p', '--pid'] },
+  setsid: {},
+  unshare: {},
+  busybox: {},
+  runuser: {
+    valueOptions: ['-u', '--user', '-g', '--group', '-G', '--supp-group'],
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Done: `-s` / `--shell` are value options of `runuser` now, so `runuser -u postgres -s /usr/local/bin/fish -c 'psql -c "drop table users"'` reaches the command-string scan. Tests for the separate and `--shell=` forms.

**Comment by L4XB** [comment]
#3107 fixed #3024 on `main` while this was open, so I restarted the branch from `main`'s code and narrowed it to what that fix still allows. `main` at `bf70150` lets six wrapper shapes reach a SQL client unseen: `bash -lc`, `env -S` and `--split-string`, and the `time`, `nice` and `timeout` runners. It also allows anything past the `sh -c` recursion limit, and five levels already hide a `drop table`. The table in the description has before/after for each. It's a small change on #3107's own functions, with tests that fail on `main`.

File: scripts/hooks/gateguard-fact-force.js [hunk 8]
```diff
@@ -487,11 +504,63 @@ function unwrapLeadWrappers(tokens) {
...
+ * @returns {string[]} the command lines, possibly none
+ */
+function envSplitStrings(tokens) {
+  const payloads = [];
+  for (let i = 0; i < tokens.length; i += 1) {
+    if (commandBasename(tokens[i]) !== 'env') continue;
+    for (let j = i + 1; j < tokens.length; j += 1) {
+      const arg = tokens[j];
+      if (arg === '-S' || arg === '--split-string') {
+        if (j + 1 < tokens.length) payloads.push(tokens.slice(j + 1).join(' '));
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Confirmed and fixed in 64e004a96: arguments after the `env -S` string are now escaped back into single literal words, so your case is allowed, while `env -S 'psql -c' 'drop table users'` is still denied. Both are new cells.

**Comment by chenhz01** [comment]
One miss I don't see in the long discussion: The runner table misses `su`. Only `sudo`, `doas`, and `env` keep dedicated unwrap handling and `su` is absent from `RUNNER_PREFIXES` — so `su postgres -c "psql -c 'drop table users'"` resolves the basename to `su` and the SQL-client check never sees `psql`. Fix shape: treat `su` like the runners — consume login flags and one user positional, then route the `-c`/`--command` value back through segment scanning:

```
if (base === 'su') { skip '-'/'-l' + one user positional; mark next -c value as its own command line }
```

`taskset -c 0` is the same class (pure runner, one value option) if you want the table complete. The stacked-runner test you added is the right harness — `sudo su postgres -c ...` should be its first case. I also do this kind of work commercially — if you'd rather have it handled end-to-end, happy to scope a fixed-price engagement.

**Comment by L4XB** [comment]
Good catch, both are in bbe5d784:

- `su`: the value of `-c`/`--command`, or of a short-option cluster ending in `c` (`-lc`), is now scanned as a command line of its own, like the value of `sh -c`. The scan continues past `--`, because `su` passes the arguments after it to the target shell, which honours a `-c` there too. Covered cases: `sudo su postgres -c ...`, `su - postgres -c ...`, `su -lc ... postgres`, `su --command=... postgres` and `su postgres -- -c ...`. `su postgres -c "psql -c 'select 1'"` stays allowed.
- `taskset`: joins the runners with one positional (the mask); its `-c` is a flag, so `taskset -c 0 psql ...` resolves to `psql`.

246 tests pass, and removing each of the five pieces fails the runner test.

File: scripts/hooks/gateguard-fact-force.js [hunk 9]
```diff
@@ -359,6 +359,25 @@ function quoteAwareSegments(input) {
...
+ *
+ * @param {string[]} tokens dequoted tokens for one segment
+ * @param {number} start index of the `su` token
+ * @returns {string|null}
+ */
+function suCommandLine(tokens, start) {
+  for (let i = start + 1; i < tokens.length; i += 1) {
+    const arg = tokens[i];
+    if (arg.startsWith('--command=')) return arg.slice('--command='.length);
+    if (arg === '--command' || /^-[A-Za-z]*c$/.test(arg)) return i + 1 < tokens.length ? tokens[i + 1] : null;
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Valid: `su` keeps the last `-c`, and only the first was scanned. Fixed in 2df1aa7a: every `-c`/`--command` value is now scanned as a command line of its own, so it does not matter which one wins. Both orders are tested (`-c "echo ok" -c "psql …drop…"` and the reverse), and scanning only the first value fails the test.

File: scripts/hooks/gateguard-fact-force.js [hunk 10]
```diff
@@ -359,6 +359,31 @@ function quoteAwareSegments(input) {
...
+  for (let i = start + 1; i < tokens.length; i += 1) {
+    const arg = tokens[i];
+    if (arg.startsWith('--command=')) {
+      commandLines.push(arg.slice('--command='.length));
+    } else if ((arg === '--command' || /^-[A-Za-z]*c$/.test(arg)) && i + 1 < tokens.length) {
+      commandLines.push(tokens[i + 1]);
+      i += 1;
+    }
+  }
+  return commandLines;
```
[the remark this answers is not in the document]
Comment (L4XB) [reply scripts/hooks/gateguard-fact-force.js]: Keeping this one deliberately. `su` runs the last `-c`, so checking every value can deny a command line whose final `-c` is harmless. But the only way to hit that is to write a destructive `-c` and then override it, which is an unlikely form for real use. Checking only the last value would make the guard depend on su's option-precedence rule matching our reading of it. For an approval gate, a rare extra prompt is the cheaper failure. If you would rather check only the effective command, it is a one-line change and I can make it.
